lib/xebow/rgb_matrix.ex

```diff
@@ -2,19 +2,16 @@ defmodule Xebow.RGBMatrix do
   use GenServer
 
   alias Circuits.SPI
-  alias Xebow.RGBMatrix.Animations
+  alias Xebow.Animation
 
   import Xebow.Utils, only: [mod: 2]
 
   defmodule State do
-    @fields [:spidev, :animation, :animation_state]
-    @enforce_keys @fields
-
-    defstruct @fields
+    defstruct [:spidev, :animation]
   end
 
-  @type pixels :: list({non_neg_integer, non_neg_integer})
-  @type colors :: list(any)
+  @type pixel :: {non_neg_integer, non_neg_integer}
+  @type pixel_color :: Chameleon.HSV.t()
```

Comment:
FYI, I converted these to the singular form of the type since it's easier to convert a singular type into a list (`list(pixel)` or `[pixel]`) than it is to go the other way around. It also simplifies the readability of the type definition a little, and pushes the visibility of a list of a given type into the specs of the functions that consume/produce a list vs. a single type.

For example, if I see this...

```ex
@spec paint(pixels :: pixels) :: pixels
```

I can assume that `pixels` is a list, but I don't really know for sure unless I check the type definition of `pixels`. Maybe it was a map with a weird name.

But like this...

```ex
@spec paint(pixels :: [pixel]) :: [pixel]
```

It's a lot more clear for me to see that as long as something gives me a `pixel`, I can stuff it in a list and use it in this function. I can pass pixels in and out of functions like this all day long without having to look up the type definition. The only reason I would need to look up the type in this case is if I needed to access any of its properties.

Reply:
completely agree